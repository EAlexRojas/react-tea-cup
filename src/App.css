.App {

}


.tree-node__row {
    display: flex;
}

.tree-node__spacer {
    min-width: 16px;
    max-width: 16px;
}
