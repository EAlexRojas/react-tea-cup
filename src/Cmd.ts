import {Dispatcher} from "./Dispatcher";

export abstract class Cmd<Msg> {

    static none<Msg>(): Cmd<Msg> {
        return new CmdNone();
    }

    abstract run(dispatch: Dispatcher<Msg>): void;

}


class CmdNone<Msg> extends Cmd<Msg> {
    run(dispatch: Dispatcher<Msg>): void {
        // it's a noop !
    }
}

export function noCmd<Model,Msg>(model:Model): [Model, Cmd<Msg>] {
    return [ model, Cmd.none() ]
}