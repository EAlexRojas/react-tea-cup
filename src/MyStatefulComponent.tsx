import {Component} from "react";
import React from "react";

export interface MyProps {
    foo: string
}

interface MyState {
    i: number
}

export class MyStatefulComponent extends Component<MyProps, MyState> {


    constructor(props: Readonly<MyProps>) {
        super(props);
        this.state = {
            i: 0
        }
    }

    render() {
        return (
            <div>
                <h3>I am stateful</h3>
                <p>
                Yo: {this.props.foo + " " + this.state.i}
                </p>
                <button onClick={() =>
                    this.setState({...this.state, i: this.state.i + 1 })
                }>
                    Click me
                </button>
            </div>
        );
    }
}
