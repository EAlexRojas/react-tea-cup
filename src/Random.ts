import {Task} from './Task'
import {Ok, Result} from "./Result";


export function random(lo: number, hi: number): Task<void, number> {
    return new RandomTask(lo, hi)
}


class RandomTask extends Task<void, number> {

    readonly lo:number;
    readonly hi:number;

    constructor(lo: number, hi: number) {
        super();
        this.lo = lo;
        this.hi = hi;
    }

    run(): Promise<Result<void, number>> {
        return Promise.resolve(Ok<void,number>(randomIntFromInterval(this.lo, this.hi)));
    }


}

function randomIntFromInterval(min: number,max: number)  {// min and max included
    return Math.floor(Math.random()*(max-min+1)+min);
}