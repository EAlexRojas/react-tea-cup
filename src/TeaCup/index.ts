export * from './Cmd'
export * from './Dispatcher'
export * from './Memo'
export * from './Program'
export * from './Random'
export * from './Result'
export * from './Task'
export * from './Sub'
export * from './Animation'
export * from './Maybe'
export * from './List'
export * from './Navigation'
export * from './Decode'
export * from './Http'