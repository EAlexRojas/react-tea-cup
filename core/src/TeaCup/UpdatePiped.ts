/*
 * MIT License
 *
 * Copyright (c) 2019 RÃ©mi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

import { Cmd } from './Cmd';

export type UpdateFunction<Model, Msg> = GenericUpdateFunction<Model, Cmd<Msg>>;

/**
 * Piped updates.
 * Useful for chaining building blocks making up the update loop.
 */
export function updatePiped<Model, Msg>(
  model: Model,
  ...updates: readonly UpdateFunction<Model, Msg>[]
): [Model, Cmd<Msg>] {
  return genericUpdatePiped(Cmd.batch, model, ...updates);
}

export type GenericUpdateFunction<Model, Cmd> = (model: Model) => [Model, Cmd];

/**
 * Generic implementation of chained updates.
 * See updatePiped().
 */
export function genericUpdatePiped<Model, Cmd>(
  batch: (cmds: readonly Cmd[]) => Cmd,
  model: Model,
  ...updates: readonly GenericUpdateFunction<Model, Cmd>[]
): [Model, Cmd] {
  const cmd0: Cmd = batch([]);
  return updates.reduce<[Model, Cmd]>(
    (acc: [Model, Cmd], update: GenericUpdateFunction<Model, Cmd>) => {
      const [model0, cmd0] = acc;
      const [model1, cmd1] = update(model0);
      return [model1, batch([cmd0, cmd1])];
    },
    [model, cmd0],
  );
}
