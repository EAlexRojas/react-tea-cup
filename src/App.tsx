import React, {Component, ReactNode} from 'react';
// import './App.css';
import {Program} from './TeaCup/Program'
import {Dispatcher} from "./TeaCup/Dispatcher";
import {Cmd, noCmd} from "./TeaCup/Cmd";
import {random} from "./TeaCup/Random";
import {Task} from "./TeaCup/Task";
import {number} from "prop-types";


interface Model {
    value: number
    enabled: boolean
}


function init() {
    return {
        value: 0,
        enabled: true
    }
}


function view(dispatch: Dispatcher<Msg>, model:Model) {
    return (
        <div>
            Value = {model.value}
            <button
                disabled={!model.enabled}
                onClick={() => dispatch({ type: "inc" }) }>
                +
            </button>
            <button
                disabled={!model.enabled}
                onClick={() => dispatch({ type: "dec"}) }>
                -
            </button>
            <button
                disabled={!model.enabled}
                onClick={() => dispatch({ type: "rand"}) }>
                random
            </button>
            <button
                disabled={!model.enabled}
                onClick={() => dispatch({ type: "timeout"}) }>
                timeout
            </button>
        </div>
    );
}


type Msg
    = Increment
    | Decrement
    | Randomize
    | RandomReceived
    | Timeout
    | TimeoutReceived


interface Increment {
    type: "inc"
}


interface Decrement {
    type: "dec"
}


interface Randomize {
    type: "rand"
}


interface RandomReceived {
    type: "random_received"
    value: number
}


interface Timeout {
    type: "timeout"
}


interface TimeoutReceived {
    type: "timeout_received"
    value: number
}


function update(msg: Msg, model: Model): [Model, Cmd<Msg>] {
    switch (msg.type) {
        case "inc":
            return noCmd({ ...model, value: model.value + 1 });
        case "dec":
            return noCmd({...model, value: model.value - 1});
        case "rand":
            const rndTask = random(0, 100);
            const cmd = Task.perform(rndTask, i => {
                return {
                    type: "random_received",
                    value: i
                } as Msg
            });
            return [
                model,
                cmd
            ];
        case "random_received":
            return noCmd({...model, value: msg.value});
        case "timeout":
            return [
                {...model, enabled: false},
                new MyTimeout(1000)
            ];
        case "timeout_received":
            return noCmd({...model, value: msg.value, enabled: true});
    }
}


class MyTimeout extends Cmd<TimeoutReceived> {

    readonly timeoutMs: number;

    constructor(timeoutMs: number) {
        super();
        this.timeoutMs = timeoutMs;
    }

    run(dispatch: Dispatcher<TimeoutReceived>): void{
        setTimeout(() => {
            dispatch({
                type:"timeout_received",
                value: new Date().getTime()
            })
        }, this.timeoutMs)
    }
}


class App extends Component {
  render() {
    return (
      <div className="App">
        <Program init={init} view={view} update={update}/>
      </div>
    );
  }
}

export default App;
