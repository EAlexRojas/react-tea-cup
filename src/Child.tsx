import {Dispatcher} from "./TeaCup/Dispatcher";
import React from 'react';
import {Cmd, noCmd} from "./TeaCup/Cmd";

export interface Model {
    value: string
}

export type Msg
    = { type: "add" }
    | { type: "remove" }


export function init(): Model {
    return {
        value: "abcdefg"
    }
}


export function view(dispatch: Dispatcher<Msg>, model: Model) {
    return (
        <div>
            <button
                disabled={model.value.length == 0}
                onClick={() => dispatch({type: "remove"})}>
                Remove char
            </button>
            <p>Text: <em>{model.value}</em></p>
            <button
                disabled={model.value.length == 0}
                onClick={() => dispatch({type: "add"})}>
                Add char
            </button>
        </div>
    )
}


export function update(msg: Msg, model: Model): [Model, Cmd<Msg>] {
    switch (msg.type) {
        case "add":
            return noCmd({...model, value: model.value + "x"});
        case "remove":
            return noCmd(
                model.value.length > 0
                    ? {...model, value: model.value.substring(0, model.value.length - 1)}
                    : model
            );
    }
}