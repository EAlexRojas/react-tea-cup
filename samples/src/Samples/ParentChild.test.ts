/*
 * MIT License
 *
 * Copyright (c) 2019 RÃ©mi Van Keisbelck
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

import { view, Msg, update, init, Model } from "./ParentChild";
import { shallow, mount } from 'enzyme';
import { Nothing, nothing, just, Cmd, Task } from "../../../src/TeaCup";


describe("Test ParentChild", () => {

    describe("init state", () => {

        test("counter model x3", async () => {
            const [state, cmd] = init();
            expect(state).toHaveLength(3);
            // TODO batched Cmd.none() untestable
        });

    });

    describe("view state", () => {

        const noop = () => { }
        const [initialState, _cmd] = init();

        test("snapshot initial", () => {
            const wrapper = mount(view(noop, initialState))
            expect(wrapper).toMatchSnapshot();
        });

        test("three counters", () => {
            const wrapper = mount(view(noop, initialState))
            expect(wrapper.find('.counter')).toHaveLength(3);
        });

    });

    describe("clicking generates messages", () => {
        var captured: Msg | undefined;
        const captureMsg = (msg: Msg) => captured = msg

        beforeEach(() => {
            captured = undefined;
        });

        const [initialState, _cmd] = init();

        test('decrement first child', () => {
            const wrapper = mount(view(captureMsg, initialState))
            wrapper.find('.counter > button').at(0).simulate('click')
            expect(captured).toEqual({
                childIndex: 0,
                childMsg: { type: 'dec' }
            })
        });
    });
});

