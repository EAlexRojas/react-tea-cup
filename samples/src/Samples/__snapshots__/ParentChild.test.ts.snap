// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ParentChild view state snapshot initial 1`] = `
<div>
  <div
    key="0"
  >
    <div
      className="counter"
    >
      <button
        onClick={[Function]}
      >
        -
      </button>
      <span>
        0
      </span>
      <button
        onClick={[Function]}
      >
        +
      </button>
    </div>
  </div>
  <div
    key="1"
  >
    <div
      className="counter"
    >
      <button
        onClick={[Function]}
      >
        -
      </button>
      <span>
        0
      </span>
      <button
        onClick={[Function]}
      >
        +
      </button>
    </div>
  </div>
  <div
    key="2"
  >
    <div
      className="counter"
    >
      <button
        onClick={[Function]}
      >
        -
      </button>
      <span>
        0
      </span>
      <button
        onClick={[Function]}
      >
        +
      </button>
    </div>
  </div>
</div>
`;
